tree-sitter/src/TreeSitter/GenerateSyntax.hs

```diff
@@ -50,82 +50,93 @@ astDeclarationsForLanguage language filePath = do
 syntaxDatatype :: Ptr TS.Language -> Datatype -> Q [Dec]
 syntaxDatatype language datatype = case datatype of
   SumType (DatatypeName datatypeName) _ subtypes -> do
-    cons <- traverse (constructorForSumChoice datatypeName) subtypes
-    result <- symbolMatchingInstanceForSums language name subtypes
-    pure $ generatedDatatype name cons:result
+    typeParameterName <- newName "a"
+    cons <- traverse (constructorForSumChoice datatypeName typeParameterName) subtypes
+    result <- symbolMatchingInstanceForSums language name subtypes typeParameterName
+    pure $ generatedDatatype name cons typeParameterName:result
   ProductType (DatatypeName datatypeName) _ children fields -> do
-    con <- ctorForProductType datatypeName children fields
-    result <- symbolMatchingInstance language name datatypeName
-    pure $ generatedDatatype name [con]:result
+    typeParameterName <- newName "a"
+    con <- ctorForProductType datatypeName typeParameterName children fields
+    result <- symbolMatchingInstance language name datatypeName typeParameterName
+    pure $ generatedDatatype name [con] typeParameterName:result
   LeafType (DatatypeName datatypeName) named -> do
-    con <- ctorForLeafType named (DatatypeName datatypeName)
-    result <- symbolMatchingInstance language name datatypeName
+    typeParameterName <- newName "a"
+    con <- ctorForLeafType named (DatatypeName datatypeName) typeParameterName
+    result <- symbolMatchingInstance language name datatypeName typeParameterName
     pure $ case named of
-      Anonymous -> generatedDatatype name [con]:result
-      Named -> NewtypeD [] name [] Nothing con deriveClause:result
+      Anonymous -> generatedDatatype name [con] typeParameterName:result
+      Named -> DataD [] name [PlainTV typeParameterName] Nothing [con] deriveClause:result
```

Comment:
Could this second case use `generatedDatatype` now?

Reply:
Yes, and it does now!